---
title: "ScRef_data"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r include=FALSE}

#removes all data from global environment
rm(list = ls())
```


## Import Libraries 
```{r librarys}
library(Seurat)
library(SeuratDisk)
library(sctransform)
library(ggplot2)
library(Matrix)
library(readr)
```


## Set Directories
```{r set_dirs}
input_dir <- "I:/GMLSC1-Q2051/Pfizer/Reference_Data/Wu_etal_2021_BRCA_scRNASeq/sc_mtx_import/filtered_features_zipped/"
output_dir <- "I:/GMLSC1-Q2051/Pfizer/Reference_Data/Wu_etal_2021_BRCA_scRNASeq/"
```


## Import scRef dataset
```{r import_data}
## Import Ref matrix
mtx <- Matrix::readMM(paste0(input_dir,"matrix.mtx.gz"))

## Import Ref barcodes
barcodes <- readr::read_tsv(paste0(input_dir,"barcodes.tsv.gz"), col_names = FALSE)

## Import Ref Features
features <- readr::read_tsv(paste0(input_dir,"features.tsv.gz"), col_names = FALSE)
```


## Create Seurat Object
```{r create_seurat_object}

colnames(mtx) <- barcodes$X1
rownames(mtx) <- features$X1

ref_data <- CreateSeuratObject(mtx)
```


## Add metadata
```{r add_metadata}
reference  <- LoadH5Seurat("I:/GMLSC1-Q2051/Pfizer/Analysis_Results/Jacky/BC_atlas_xe.h5seurat")

### Match barcodes and metadata


reduced_ref_data <- subset(ref_data, cells = colnames(reference))

for (column in colnames(reference@meta.data)){
  reduced_ref_data <- AddMetaData(object = reduced_ref_data, metadata = reference@meta.data[column],col.name = column)
}

```


## Process Data
```{r process_data}
sc_pipeline <- function(sc_data) {
  
  # store mitochondrial percentage in object meta data
  sc_data <- PercentageFeatureSet(sc_data, pattern = "^MT-", col.name = "percent.mt")
  
  sc_data_removed_cells <- subset(sc_data, subset = nFeature_RNA > 20)
  selected_genes <-rownames(sc_data)[Matrix::rowSums(sc_data@assays$RNA@layers$counts)>3] #removes genes expressed in <4 cells
  sc_data_removed_genes <-subset(sc_data_removed_cells, features = selected_genes)
  

  # run sctransform
  SCT_data <- SCTransform(sc_data_removed_genes, method = "glmGamPoi",verbose = FALSE, assay = "RNA")
  
  # These are now standard steps in the Seurat workflow for visualization and clustering
  SCT_data <- RunPCA(SCT_data, verbose = FALSE)
  SCT_data <- RunUMAP(SCT_data, dims = 1:30, verbose = FALSE)
  
  SCT_data <- FindNeighbors(SCT_data, dims = 1:30, verbose = FALSE)
  SCT_data <- FindClusters(SCT_data, verbose = FALSE)
  
  #Saves processed single-cell dataset 
  saveRDS(SCT_data, file = paste0(output_dir,"Ref_data_processed.RDS"))
  
  return(SCT_data)
}

processed_ref <- sc_pipeline(reduced_ref_data)
```

## Plot Data
```{r process_data}
DimPlot(processed_ref, group.by = "celltype_major")
DimPlot(processed_ref, group.by = "celltype_minor")

```