---
title: "Pfizer_single_cell"
output: html_document
date: "2023-09-06"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r include=FALSE}

#removes all data from global environment
rm(list = ls())
```


## Import Libraries 
```{r librarys}
library(Seurat)
library(ggplot2)
library(dplyr)
library(scCustomize)
library(pheatmap)
library(tidyr)
library(gridExtra)
```


## Setup Working Directories
```{r wd}
input_dir <- "I:/GMLSC1-Q2051/Pfizer/scRNAseq/SN2_Pfizer_per_sample_outs/"
output_dir <- "I:/GML001-Q1851/Andrew_C/Pfizer/scRNA/"
```


## Setup Data and Processing Pipeline
```{r sample_processing}

sc_pipeline <- function(ID, inputdir, outputdir) {
  
  outdir <- paste0(outputdir,ID,"/")
  dir.create(outdir)
  
  sc_data.data <- Read10X(data.dir = paste0(inputdir,ID,"/count/sample_filtered_feature_bc_matrix"))
  # Initialize the Seurat object with the raw (non-normalized data).
  sc_data <- CreateSeuratObject(counts = sc_data.data)
  
  # store mitochondrial percentage in object meta data
  sc_data <- PercentageFeatureSet(sc_data, pattern = "^MT-", col.name = "percent.mt")
  
  # QC
  VlnPlot(sc_data, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
  #ggsave(paste0(outdir,ID,"_QC_Vln.png"), height=8, width=16)
  
  selected_c <- WhichCells(sc_data, expression = nFeature_RNA > 20)
  selected_f <- rownames(sc_data)[Matrix::rowSums(sc_data) > 3]

  sc_data <- subset(sc_data, features = selected_f, cells = selected_c)
  
  
  # run sctransform
  sc_data <- SCTransform(sc_data, verbose = FALSE)
  
  # These are now standard steps in the Seurat workflow for visualization and clustering
  sc_data <- RunPCA(sc_data, verbose = FALSE)
  sc_data <- RunUMAP(sc_data, dims = 1:30, verbose = FALSE)
  
  sc_data <- FindNeighbors(sc_data, dims = 1:30, verbose = FALSE)
  
  return(sc_data)
}

sc_cluster <- function(sc_data, ID, outputdir, res = 0.5){
  
  #outdir <- paste0(outputdir,ID,"/")
  #dir.create(outdir)
  
  sc_data <- FindClusters(sc_data, res = res, verbose = FALSE)
  
  #Saves processed single-cell dataset 
  #saveRDS(sc_data, file = paste0(outdir, ID, "_scRNA_seurat.RDS"))
  return(sc_data)
}

``` 



## Run samples
```{r run_samples}

input_sample_list <- c("AVD_61FEX_1302A","AVD_61FEX_1346A","AVD_61FEX_1956A","AVD_61FEX_2116A","AVD_61FEX_3561A","AVD_61FEX_3633A","AVD_61FEX_4851A","AVD_61FEX_5626A","AVD_61FEX_6481A","AVD_61FEX_6841A","AVD_61FEX_9982A")
sample_list <-c("1302A","1346A","1956A","2116A","3561A","3633A","4851A","5626A","6481A","6841A","9982A")

data_list <- list()

for (sample in input_sample_list){
  data <- sc_pipeline(sample, inputdir = input_dir, outputdir = output_dir)
  data_list[[sample]] <- data
}

```


```{r}

for (sample in names(data_list)){
  data <- sc_cluster(data_list[[sample]],ID = sample, outputdir = output_dir, res = 0.5)
  data_list[[sample]] <- data
}


dim_plots <- list()

for(sample in names(data_list)){
  dim_plots[[sample]]<- DimPlot(data_list[[sample]], group.by = "seurat_clusters")
}

do.call("grid.arrange", dim_plots)

```








## Import scReference data
```{r import_ref_data}

##Create Reference directory
outdir_ref <- paste0(output_dir,"Reference_Data/")
#dir.create(outdir_ref)


####load References dataset
#reference  <- readRDS(file = "I:/GMLSC1-Q2051/Pfizer/Reference_Data/Wu_etal_2021_BRCA_scRNASeq/Ref_data_processed.RDS")

## Subset to ER+ (LumA&LumB)
#reference_processed <- subset(reference, subset = subtype == "ER+")

## Run label transfer
#reference_processed <- reference_processed[,unname(which(colSums(GetAssayData(reference_processed))!=0))]

#Ref_data <- reference_processed

#Idents(Ref_data) <- "celltype_major"
#Ref_data_subset <- subset(Ref_data, idents = "Normal Epithelial", invert = TRUE)

#saveRDS(Ref_data_subset, file = paste0(outdir_ref, "Reference_scRNA_seurat.RDS"))
Ref_data_subset <- readRDS(paste0(outdir_ref, "Reference_scRNA_seurat.RDS"))

```


## Run Label Transfer
```{r label_transfer}
run_label_transfer <- function(Ref_data, Query_data, ID, output_dir) {
 
  outdir <- paste0(output_dir,ID,"/")
  
  comm_gene1 <- intersect(rownames(Ref_data), 
                        rownames(Query_data))
  
  anchors <- FindTransferAnchors(reference = Ref_data, 
                                 query = Query_data, features = comm_gene1)
  sc_obj.predictions.assay <- TransferData(anchorset = anchors, 
                                              refdata = Ref_data$celltype_major,  
                                              weight.reduction = Query_data[["pca"]], 
                                              dims = 1:40) 
  sc_obj.predictions.assay1 <- TransferData(anchorset = anchors, 
                                           refdata = Ref_data$celltype_minor,  
                                           weight.reduction = Query_data[["pca"]], 
                                           dims = 1:40) 
  rownames(sc_obj.predictions.assay) <- colnames(Query_data)
  rownames(sc_obj.predictions.assay1) <- colnames(Query_data)
  
  labels <- cbind(sc_obj.predictions.assay,sc_obj.predictions.assay1)
  #write.csv(labels, file=paste0(outdir, ID,"_label_transfer.csv"))
  
  Query_data$celltype_major <- sc_obj.predictions.assay$predicted.id
  Query_data$celltype_minor <- sc_obj.predictions.assay1$predicted.id
    
  return(Query_data)
}

```



```{r add_labels}

lbtr_list <- list()

for (sample in names(data_list)){
  data <- run_label_transfer(Ref_data_subset, data_list[[sample]], sample,output_dir = output_dir)
  lbtr_list[[sample]] <- data
}


idx <- 1
while(idx <= length(sample_list)){
  saveRDS(lbtr_list[[idx]], file = paste0(output_dir, input_sample_list[idx],"/", input_sample_list[idx],"_scRNA_label_transfer.RDS"))
  idx <- idx + 1
}

saveRDS(lbtr_list, file = paste0(output_dir, "all_scRNA_label_transfer.RDS"))
```


```{r load_saved_data}
lbtr_list <- readRDS(file = paste0(output_dir, "all_scRNA_label_transfer.RDS"))
```






```{r}
plots <- list()

for(i in seq(1,length(lbtr_list))){
  plot <- DimPlot(lbtr_list[[i]],group.by = "seurat_clusters")
  plots[[i]] <- plot
}

do.call("grid.arrange", c(plots, ncol=3))


lbtr_plots <- list()

for(i in seq(1,length(lbtr_list))){
  lbtr_plots[[i]]<- DimPlot(lbtr_list[[i]], group.by = "celltype_major")
}

do.call("grid.arrange", lbtr_plots)



```


```{r}
#df <- read.csv("/Users/uqacause/Downloads/41467_2022_34581_MOESM4_ESM.csv")
markers <- read.csv("I:/GML001-Q1851/Andrew_C/Pfizer/metastatic_breast_cancer_reference_marker_genes.csv")

marker_genes <- split(markers$gene, markers$cluster)

unique_genes <- unlist(marker_genes)
genes_count <- table(unique_genes)

# Get genes that appear only once
single_occurrence_genes <- names(genes_count[genes_count == 1])

# Filter original list to keep only genes that appear once
filtered_list <- lapply(marker_genes, function(cluster_genes) {
  cluster_genes[cluster_genes %in% single_occurrence_genes]
})

# Get the first 5 genes for each key
top_5 <- lapply(filtered_list, function(cluster_genes) {
  head(cluster_genes, n = 5)
})



heatmaps <- list()

for(i in seq(1,length(lbtr_list))){
  plot <- DotPlot(lbtr_list[[i]], features = top_5 ,group.by = "celltype_major") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.text.x.top = element_text(angle = 90, vjust = 0.5, hjust=1))+NoLegend()
  heatmaps[[sample_list[[i]]]] <- plot

}

plot <- do.call("grid.arrange", c(heatmaps, ncol=3))






color_dict <- list(
  "Cancer Epithelial"="#b3b1a6",
  "Plasmablasts"= "#24d7ff",     
  "B-cells" = "#33A02C",      
  "T-cells" =  "#E31A1C",     
  "CAFs" = "#FB9A99",          
  "Myeloid" =  "#1F78B4" ,        
  "Endothelial"= "#E8d9fc",
  "PVL"= "#FF7F00"                  
  )



plot_annotations <- function(data, sample, output_dir, adding_graph = FALSE){
  
  plot1 <- DimPlot(data, group.by = "seurat_clusters")
  plot2 <- DimPlot(data, group.by = "celltype_major", cols = color_dict)
  
  cell_type_counts <- data@meta.data %>%
    group_by(seurat_clusters, celltype_major) %>%
    summarise(count = n())
  
 cell_type_counts <- cell_type_counts %>%
  mutate(color = unlist(color_dict[celltype_major]))
 
 unique_labels <- unique(cell_type_counts$celltype_major)
  
 # Create a named vector of corresponding colors
 legend_colors <- unlist(color_dict[unique_labels])

 plot3 <- ggplot(cell_type_counts, aes(y = count, x = seurat_clusters, fill = celltype_major, color = celltype_major)) +
    geom_bar(position = "fill", stat = "identity", width = 0.8) +
    scale_fill_manual(values = unlist(color_dict)) +
    scale_color_manual(values = unlist(color_dict)) +
    coord_flip() + theme_classic() + labs(x = "Cluster", y = "Proportion")
  
 
 cell_types <- data@meta.data %>%
    group_by(celltype_major) %>%
    summarise(count = n()) %>%
    mutate(Proportion = count / sum(count))
 
 cell_types$samples <- sample
 
 plot5 <- DotPlot(data, features = top_5 ,group.by = "celltype_major") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+NoLegend()

  plot4 <- DotPlot(data, features = top_5 ,group.by = "seurat_clusters") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+NoLegend()
  
  
  plot <- (plot1 + plot2+plot3)/(plot4+plot5)
  ggsave(paste0(output_dir, sample, "/", sample,"_new_annotations.pdf"),plot, height = 10, width = 20)
 
  if (adding_graph == TRUE){
    return(plot)
  }
  
}

plot_list <- list()

idx <- 1
while (idx <= length(input_sample_list)){
  plot <- plot_annotations(lbtr_list[[idx]], input_sample_list[idx],output_dir = output_dir, adding_graph = TRUE)
  plot_list[[idx]] <- plot
  idx <- idx+1
}

```


















## Plot label transfer umaps and cell type proportions
```{r plot_umaps}

color_dict <- list(
  "Cancer Epithelial"="#b3b1a6",
  "Plasmablasts"= "#24d7ff",     
  "B-cells" = "#33A02C",      
  "T-cells" =  "#E31A1C",     
  "CAFs" = "#FB9A99",          
  "Myeloid" =  "#1F78B4" ,        
  "Endothelial"= "#E8d9fc",
  "PVL"= "#FF7F00"                  
  )



plot_umap <- function(data_og, data, sample, output_dir, adding_graph = FALSE){
  
  plot1 <- DimPlot(data_og)
  plot2 <- DimPlot(data, group.by = "celltype_major", cols = color_dict)
  
  cell_type_counts <- data@meta.data %>%
    group_by(seurat_clusters, celltype_major) %>%
    summarise(count = n())
  
 cell_type_counts <- cell_type_counts %>%
  mutate(color = unlist(color_dict[celltype_major]))
 
 unique_labels <- unique(cell_type_counts$celltype_major)
  
 # Create a named vector of corresponding colors
 legend_colors <- unlist(color_dict[unique_labels])

 plot3 <- ggplot(cell_type_counts, aes(y = count, x = seurat_clusters, fill = celltype_major, color = celltype_major)) +
    geom_bar(position = "fill", stat = "identity", width = 0.8) +
    scale_fill_manual(values = unlist(color_dict)) +
    scale_color_manual(values = unlist(color_dict)) +
    coord_flip() + theme_classic() + labs(x = "Cluster", y = "Proportion")
  
 
 cell_types <- data@meta.data %>%
    group_by(celltype_major) %>%
    summarise(count = n()) %>%
    mutate(Proportion = count / sum(count))
 
 cell_types$samples <- sample
 
 plot4 <- ggplot(cell_types, aes(y = count, x = samples, fill = celltype_major  , color = celltype_major)) +
    geom_bar(position = "fill", stat = "identity", width = 0.8) +
    scale_fill_manual(values = unlist(color_dict)) +
    scale_color_manual(values = unlist(color_dict)) + theme_classic() + labs(x = "Cluster", y = "Proportion")
  
  
  plot <- (plot1 + plot2)/(plot3 + plot4)
  ggsave(paste0(output_dir, sample, "/", sample,"_umap_plots.pdf"),plot, height = 15, width = 20)
 
  if (adding_graph == TRUE){
    return(plot)
  }
  
}

plot_list <- list()

idx <- 1
while (idx <= length(input_sample_list)){
  plot <- plot_umap(data_list[[idx]], lbtr_list[[idx]], input_sample_list[idx],output_dir = output_dir, adding_graph = TRUE)
  plot_list[[idx]] <- plot
  idx <- idx+1
}

```


```{r cluster_annotation}

celltype_colours <- list(
  "Luminal_Epithelial"="#b3b1a6",
  "Plasma"= "#24d7ff",     
  "B_cells" = "#33A02C",      
  "T_cells" =  "#E31A1C",     
  "CAFs" = "#FB9A99",          
  "Myeloid" =  "#1F78B4" ,        
  "Endothelial"= "#E8d9fc",
  "Myoepithelial"= "#FF7F00",
  "Fibroblast" = "yellow",
  "Adipocyte" = "#B2DF8A"
  )


breast_genes_cellxgene <- list(
  "T_cells"= c("CCL5", "IL7R", "CD2", "LEPROTL1", "SARAF"),
  "B_cells"= c("CD37", "IGHM", "MS4A1"),
  "Plasma"= c("SSR4", "JCHAIN", "SEC11C", "MZB1", "IGHA1", "HERPUD1"),
  "Myeloid" = c("LYZ", "HLA-DPB1","HLA-DPA1", "HLA-DQA1","HLA-DQB1","HLA-DRA"),
  "Myoepithelial"= c("KRT17", "KRT14", "DST", "KRT5", "MIR205HG"),
  "Adipocyte"= c("ACACB", "SOX5", "PDE3B", "EBF1", "WDPCP"),
  "Endothelial"= c("IFI27", "ADAMTS9", "CLDN5", "TM4SF1"),
  "Fibroblast"= c("DCN", "LUM", "APOD", "TNFAIP6","DCLK1","PTPRG", "DLC1"),
  "Luminal_Epithelial"= c("AZGP1", "KRT18", "KRT19", "CHASERR", "CD24")
  )

breast_genes_chatGPT <- list(
  "T_cells"= c("CD3E", "CD4", "CD8A", "CD45", "CCR7"),
  "B_cells"= c("CD19", "CD20", "CD21","CD79A"),
  "Plasma"= c("CD138", "IRF4", "CD38"),
  "Myeloid" = c("CD11B", "CD14","CD68", "HLA-DRA"),
  "Adipocyte"= c("FABP4", "LPL", "ACRP30", "PLIN1", "LEP","DGAT2"),
  "Endothelial"= c("PECAM1", "CD105", "CD144", "VWF", "ICAM1"),
  "Fibroblast"= c("THY1", "COL1A1", "FSP1", "VIM","ACTA2"),
  "Luminal_Epithelial"= c("ESR1", "PGR", "GATA3", "CK8")
  )
  
  
pathologist_genes <- list("Plasmablasts" = c("NUCB2","FNDC3B","XBP1","FKBP11","ANKRD28","CD38"))



#DoHeatmap(object = lbtr_list[[1]], features = breast_genes_chatGPT)
```



```{r heatmap}

heatmaps <- list()
idx <- 1
while (idx <= length(input_sample_list)){
  plot <- DotPlot(lbtr_list[[idx]], features = breast_genes_chatGPT,group.by = "seurat_clusters") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+NoLegend()
  heatmaps[[input_sample_list[[idx]]]] <- plot
  idx <- idx+1
}

do.call("grid.arrange", c(heatmaps, ncol=3))


heatmaps_cellXgene <- list()
idx <- 1
while (idx <= length(input_sample_list)){
  plot <- DotPlot(lbtr_list[[idx]], features = breast_genes_cellxgene,group.by = "seurat_clusters") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+NoLegend()
  heatmaps_cellXgene[[input_sample_list[[idx]]]] <- plot
  idx <- idx+1
}

do.call("grid.arrange", c(heatmaps_cellXgene, ncol=3))


```





```{r}

doSeuratHeatmap <- function(data, ident, even_clusters = TRUE, gene_set_dictionary = NULL,feature_list = NULL, assay = "SCT", slot = "scale.data",cells = NULL, scale = FALSE, gene.color.palatte = NULL, cluster.color.palatte = NULL, DEG.File = NULL){
  
  DefaultAssay(data) <- assay
  
  if (even_clusters == TRUE){
        cells_splt <- lapply(unique(data@meta.data[ident][,1]), function(i){
            Idents(data) <- ident
            spt <- subset(data, ident = i)
            min_cells <- min(table(data@meta.data[[ident]]))
            spt_barcodes <- sample(colnames(spt), size = min_cells, replace = FALSE)
        })
        cells <- unlist(cells_splt)
  }
  
  if (!(is.null(gene_set_dictionary))){
    features <- unique(unlist(gene_set_dictionary))
    
  } else if (!(is.null(feature_list))){
    features <- feature_list
    
  } else if (!(is.null(DEG.File))){
    top10 <- read.table(file = DEG.File)
    gene_list_from_file <- split(top10$gene, top10$cluster)
    features <- unique(unlist(gene_list_from_file))
    
  }else{
    print("ERROR: No genes were specified")
  }
  
  filtered_features <- c()
  for (feature in features){
    
    if (tryCatch({
      isFALSE(is.null(FetchData(test, c(feature), layer = slot)))
      }, error = function(e) {
        return(FALSE)
      })){
      filtered_features <- c(filtered_features, feature)
      }
  }
  
  hm <- DoHeatmap(data, features = c(filtered_features), assay = assay, slot =slot, group.bar = TRUE, cells = cells)
  
  mt_df <- hm$data
  mt_df <- mt_df %>% na.omit()
  df.wide <- pivot_wider(mt_df, names_from = Feature, values_from = Expression)
  df.wide <- as.data.frame(df.wide)
  rownames(df.wide) <- df.wide$Cell
  df.wide <-  df.wide[,-1]
  df.wide <- mutate_all(df.wide, function(x) as.numeric(as.character(x)))
  df.wide <- df.wide %>% arrange(Identity)
  
  annot_row <- df.wide %>% select(Identity)
  
  
  prev_ident <- annot_row$Identity[1]
  row_gaps <- c()
  current_change <- 0
    
  for (ident in annot_row$Identity){
      
    if(!(ident == prev_ident)){
      prev_ident <- ident
      row_gaps <- c(row_gaps, current_change)
      }
      current_change <- current_change + 1
    }
  
  
  df.wide$Identity <- NULL
  
  if (!(is.null(gene_set_dictionary))){
    gene_dict <- gene_set_dictionary
    
  } else if (!(is.null(DEG.File))){
    gene_dict <- gene_list_from_file
    
  } else {
    gene_dict <- NULL
  }
  
  if (!(is.null(gene_dict))){ 
    cell_type_genes <- c()
    for (gene in colnames(df.wide)){
      matching_keys <- sapply(gene_dict, function(genes) gene %in% genes)
      key_with_gene <- names(gene_dict[matching_keys])[1]
      cell_type_genes <- c(cell_type_genes, key_with_gene)
    }
    annot_col <- data.frame(Cell_Type = factor(cell_type_genes))
    rownames(annot_col) <- colnames(df.wide)
    
    print(annot_col)
    
    
    prev_word <- annot_col$Cell_Type[1]
    col_gaps <- c()
    current_change <- 0
    
    for (word in annot_col$Cell_Type){
      print(word)
      if(!(word == prev_word)){
        prev_word <- word
        col_gaps <- c(col_gaps, current_change)
      }
      current_change <- current_change + 1
    }

    
  } else{
    annot_col <- NULL
  }
  
  if (scale){
    scale_by <- "column"
  } else{
    scale_by <- "none"
  }
  
  
  color_pal <- list()
  
  if (!(is.null(cluster.color.palatte))){
    color_pal[[ident]] <- unlist(cluster.color.palatte)
  }
  else if (!(is.null(gene.color.palatte))){
    color_pal[["Cell_Type"]] <- unlist(gene.color.palatte)
  } else {
    color_pal <- NULL
  }
 
  
  pht <- pheatmap(df.wide, cluster_cols = FALSE, cluster_rows = FALSE, annotation_row = annot_row, gaps_row = row_gaps, show_rownames = FALSE, angle = "90", annotation_col = NULL, gaps_col = col_gaps, annotation_legend = TRUE, scale = scale_by, annotation_colors = color_pal)
  
  return(list(hm,pht))
}

```

```{r}

heatmaps <- list()
idx <- 1
while (idx <= length(input_sample_list)){
  plots <- doSeuratHeatmap(lbtr_list[[idx]], gene_set_dictionary = breast_genes_chatGPT, "seurat_clusters")
  heatmaps[[input_sample_list[idx]]] <- plots
  idx <- idx+1
}

#test_out <- doSeuratHeatmap(test, DEG.File = paste0(output_dir,sample,"/DE_analysis/", sample,"_pos_top10markers.txt"), "seurat_clusters")

```


## Refine Celltyping by clusters
```{r refine_celltyping}

get_cluster_genes <- function(data, sample, outdir){
  
  ##Make New Directory for Analsysis files
  outdir.DE <- paste0(outdir,sample,"/DE_analysis/")
  dir.create(outdir.DE)

  Idents(data) <- "seurat_clusters" 

  #### Look at only UP-regulated Genes ####
  markers <- FindAllMarkers(data, assay = "SCT", slot = "data", only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
  
  # filter to only significant hits
  markers <- markers %>% filter(p_val_adj <= 0.0001)
  #write.table(markers, file = paste0(outdir.DE, sample,"_pos_allmarkers.txt"), sep = "\t", quote = FALSE, col.names = NA)

  # Gets Top10 markers for each cluster
  #top10 <- markers %>% group_by(cluster) %>% arrange(-avg_log2FC) %>% dplyr::slice(1:10)
  #write.table(top10, file = paste0(outdir.DE, sample,"_pos_top10markers.txt"), sep = "\t", quote = FALSE, col.names = NA)
  
  return(top10)

}

All_Sample_Marker_Gene_Dictionary <- list()

idx <- 1
while (idx <= length(input_sample_list)){
  marker_gene_df <- get_cluster_genes(lbtr_list[[idx]], input_sample_list[idx],output_dir)
  All_Sample_Marker_Gene_Dictionary <- c(All_Sample_Marker_Gene_Dictionary, setNames(list(marker_gene_df), input_sample_list[idx]))
  
  idx <- idx+1
}

```




```{r enrichr}
#All_Sample_Marker_Gene_Dictionary



#dbs <- c("HuBMAP_ASCTplusB_augmented_2022","PanglaoDB_Augmented_2021","Jensen_TISSUES","GTEx_Tissue_Expression_Up","Human_Gene_Atlas")

#db <- c("HuBMAP_ASCTplusB_augmented_2022")

#for (sample in names(All_Sample_Marker_Gene_Dictionary)){
#  data <- All_Sample_Marker_Gene_Dictionary[[sample]]
#  data_dic <- split(data$gene, data$cluster)
  
#  enrichr_results <- list()
#  for(cluster in names(data_dic)){
#    enriched <- enrichr(data_dic[[cluster]], db)
#    enrichr_results[[cluster]] <- enriched
#    
#    print(plotEnrich(enriched[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value"))
#  }
  
  
#}

```






## Plot label transfer umaps and cell type proportions for reference dataset
```{r ref_dataset_plot}

####plot reference

plot1 <- DimPlot(Ref_data_subset, group.by = "celltype_major", cols = color_dict)

cell_types <- Ref_data_subset@meta.data %>%
    group_by(celltype_major) %>%
    summarise(count = n()) %>%
    mutate(Proportion = count / sum(count))

cell_types$samples <- "Reference"
 
plot2 <- ggplot(cell_types, aes(y = count, x = samples, fill = celltype_major  , color = celltype_major)) +
    geom_bar(position = "fill", stat = "identity", width = 0.8) +
    scale_fill_manual(values = unlist(color_dict)) +
    scale_color_manual(values = unlist(color_dict)) + theme_classic() + labs(x = "Cluster", y = "Proportion")
  
ggsave(paste0(outdir_ref,"Reference_umap_plots.pdf"), plot1+plot2, height = 8, width = 15) 
plot1+plot2

```


## Plot Marker Genes for cell types 
```{r check_markers}

marker_features <- c("CD3E","CD4","CD3D", #T cells
                     "CD79A","CD79B", "CD22", #B cells
                     "MZB1","ITM2C", #Plasmoblasts
                     "LYZ","CD68","TREM2", #Monocytes/Meyloid
                     "MYH11","CD36", #PVLs
                     "CCND1","ESR1","GATA3", "ERBB2", # Tumour
                     "PECAM1","VWF", #Endothealial
                     "COL1A1", "COL3A1" #CAFs
                     )


plot_marker_genes <- function(data, sample, output_dir, adding_graph = FALSE){

  plot1 <- FeaturePlot(data, features = marker_features)
  plot2 <- DimPlot(data, group.by = "celltype_major", cols = color_dict)
  plot3 <- DotPlot(data, features = marker_features, group.by = "celltype_major",idents = Idents(data))+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  ggsave(paste0(output_dir, sample,"/", sample, "_marker_umaps.pdf"), plot1+plot2, height = 15, width = 20)
  ggsave(paste0(output_dir, sample,"/",sample,"_marker_plots.pdf"), plot3)
  
  if (adding_graph == TRUE){
    return(list((plot1+plot2),plot3))
  }
  
}


marker_plots <- list()
dot_plots <- list()

idx <- 1
while (idx <= length(input_sample_list)){
  plots <- plot_marker_genes(lbtr_list[[idx]], input_sample_list[idx],output_dir = output_dir, adding_graph = TRUE)
  marker_plots[[idx]] <- plots[[1]]
  dot_plots[[idx]]<- plots[[2]]
  idx <- idx+1
}


#Run plots for reference dataset
plot_marker_genes(Ref_data_subset, "Reference_Data", output_dir)
```


## Add Sample Metadata
```{r add_metadata}
metadata <- read.csv("I:/GMLSC1-Q2051/Pfizer/Annotation_from_Pfizer/Pfizer_sample_metadata.csv")

idx <- 1
while (idx <= length(input_sample_list)){
  
  #converts to same format as metadata
  new_sample_id  <- gsub("_", "-", input_sample_list[idx])
  
  #gets row in metadata
  meta_idx <- which(metadata$`Ã¯..Subject.ID` == new_sample_id)
  
  #Add metadata
  lbtr_list[[idx]]@meta.data$sample <- input_sample_list[idx]
  lbtr_list[[idx]]@meta.data$Tissue.Type <- metadata$Tissue.Type[meta_idx]
  lbtr_list[[idx]]@meta.data$Lum.Class <- metadata$Lum.Class[meta_idx]
  lbtr_list[[idx]]@meta.data$Tissue.Region <- metadata$Tissue.Region[meta_idx]
  lbtr_list[[idx]]@meta.data$Visium.Code <- metadata$Visium.Code[meta_idx]

  idx <- idx+1
}

#Saves list of scRNA objects for visium lable transfer
saveRDS(lbtr_list, file = paste0(output_dir, "scRNA_list.RDS"))


```


## Merge scRNA datasets
```{r merge_data}

scRNA.combined <- Merge_Seurat_List(lbtr_list, add.cell.ids = input_sample_list)


#Plot merged Data
DefaultAssay(scRNA.combined) <- "RNA"
scRNA.combined <- FindVariableFeatures(scRNA.combined, selection.method = "vst")
scRNA.combined <- ScaleData(scRNA.combined)
scRNA.combined <- RunPCA(scRNA.combined, verbose = TRUE, npcs=40)
scRNA.combined <- FindNeighbors(scRNA.combined, reduction = "pca", dims = 1:40)
#scRNA.combined <- FindClusters(scRNA.combined, verbose = TRUE, resolution = 0.5)
scRNA.combined <- RunUMAP(scRNA.combined, reduction = "pca", dims = 1:40)


plot1 <- DimPlot(scRNA.combined, group.by = "sample")
plot2 <- DimPlot(scRNA.combined, group.by = "Tissue.Type")
plot3 <- DimPlot(scRNA.combined, group.by = "Lum.Class")
plot4 <- DimPlot(scRNA.combined, group.by = "celltype_major", cols = color_dict)
plot5 <- DimPlot(scRNA.combined, group.by = "celltype_minor")

plots <- (plot1+plot2+plot3)/(plot4+plot5)
ggsave(paste0(output_dir, "combined_samples_umaps.pdf"),plots, height = 12, width = 20)

###Plots just cancer epithelial subtypes
Idents(scRNA.combined) <- "celltype_major"
cancer_subset <- subset(scRNA.combined, idents = "Cancer Epithelial")
plot6 <-  DimPlot(cancer_subset, group.by = "celltype_minor")

```


## Annotation of Plasmablasts vs B-cells
```{r plasmoblasts_annotation}
Plasma_vs_bcell_features <- c("NUCB2","FNDC3B","XBP1","FKBP11","ANKRD28","CD38","IGKC","MZB1","IGHG3","TNFRSF17","IGHG1","IGLV3-1","IGHG4","IGLL5","PIM2","ITM2C",
"SDC1","CAV1","SKAMF7","GAS6","PRDM1","HSPA1B","AQP3","ERN1","TRIB1","DUSP5","CD63","NEAT1")


idx <- 1
while (idx <= length(input_sample_list)){
  plot <- DotPlot(lbtr_list[[idx]], features = Plasma_vs_bcell_features, group.by = "celltype_major")+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  #ggsave(paste0(OUT_DIR,input_sample_list[idx],"_celltypes_subset.pdf"), plot)
  print(plot)
  idx <- idx + 1
}
```


# Save finalised Objects
```{r save_finalised_objects}
outdir <- "I:/GML001-Q1851/Andrew_C/Pfizer/Finalised_Objects"
dir.create(outdir)

scRNA_dir <- paste0(outdir,"/scRNA")
dir.create(scRNA_dir)

Ref_dir <- paste0(outdir,"/Reference_Data")
dir.create(Ref_dir)


#### Reference Data
saveRDS(Ref_data_subset, file = paste0(Ref_dir,"/Finalised_Reference_Dataset.RDS"))
df <- as.data.frame(Ref_data_subset@meta.data[c("celltype_major","celltype_minor")])
write.csv(df, file=paste0(Ref_dir, "/Finalsed_Reference_Dataset_Metadata.csv"))

#### Single Cell Data
saveRDS(lbtr_list, paste0(scRNA_dir,"/Finalised_scRNA_Dataset.RDS"))

idx <- 1
while (idx <= length(input_sample_list)){
  saveRDS(lbtr_list[[idx]], file = paste0(scRNA_dir,"/Finalised_",input_sample_list[idx],"_Object.RDS"))
  df <- as.data.frame(lbtr_list[[idx]]@meta.data[c("seurat_clusters","celltype_major","celltype_minor",  "sample", "Tissue.Type","Lum.Class", "Tissue.Region", "Visium.Code")])
  write.csv(df, file=paste0(scRNA_dir, "/Finalised_",input_sample_list[idx],"_Metadata.csv"))

  idx <- idx + 1
}
```




## compare annotations
```{r compare}

df <- read.csv("I:/GML001-Q1851/Andrew_C/Pfizer/brain_UniCellDeconvolution_scRNA.csv")

names(lbtr_list) <- sample_list

brain <- list(lbtr_list[[2]],lbtr_list[[3]])
names(brain) <- c(sample_list[2],sample_list[3])

check <- list()
for (sample in names(brain)){
  sub <- df[df["sample"] == sample,]
  sub$X_new <- sub("-*.$", "", sub$X)
  rownames(sub)<-sub$X_new
  brain[[sample]]@meta.data$pred_celltype_ucdbase <- sub$pred_celltype_ucdbase
  check[[sample]] <- brain[[sample]]
}








```









